File: java/cto/ITResource.java
package cto;

/**
 * @author Dan
 */
public class ITResource
{
	private double infrastructure = 2, features = 1;
	private int revenue = 1000;
	private int users = 0;
	private int turnsSinceLastCrash = 0;




	public boolean willCrashAt (double roll) {
		return getFIR() > roll && getUsers() > 0;
	}

	public double getFIR() {
		return features / infrastructure;
	}
	public int getChanceOfCrash() {
		return (int)Math.round(getFIR() * 100);
	}
	public int getNewFeaturesCost() {
		return (int)(500 * getFIR());
	}
	public int getInfrastructureImprovementCost() {
		return Math.max(800,  (int)(infrastructure / 3));
	}

	public int getCrashCost() {
		int exposure = (int)(getUsers() / 10000);
		if (exposure < 1)
			exposure = 1;
		return (int)(exposure * getUsers() * getFIR());
	}

	private double getUserComfortFactor () {
		return Math.max((1 + getTurnsSinceLastCrash()), 10) / 10;
	}

	public int getNewUsers (boolean addedFeatures) {
		int buzzDraw = addedFeatures ? 1000 : 100;
		return (int)(buzzDraw * features * getUserComfortFactor());
	}
	public int getNewRevenue (int newUsers) {
		return (int)(getUsers() - newUsers * .75) + (newUsers * 2);
	}

	private double getFeatures() {
		return features;
	}
	private double getInfrastructure() {
		return infrastructure;
	}
	public int getRevenue() {
		return revenue;
	}
	public int getTurnsSinceLastCrash() {
		return turnsSinceLastCrash;
	}
	public int getUsers() {
		return users;
	}



	public void addNewFeatures() {
		features++;
		revenue -= getNewFeaturesCost();
	}
	public void updateInfrastructure() {
		infrastructure++;
		revenue -= getInfrastructureImprovementCost();
	}

	public void incrementTurnsSinceLastCrash () {
		this.turnsSinceLastCrash++;
	}
	public void resetTurnsSinceLastCrash() {
		this.turnsSinceLastCrash = -1;
	}

	public void addUsers (int newUsers) {
		this.users += newUsers;
	}
	public void addRevenue (int newRevenue) {
		this.revenue += newRevenue;
	}

}

File: ruby/play_the_cto/lib/play_the_cto/it_resource.rb
module PlayTheCto

  class ITResource
    def initialize
      @infrastructure = 2.0
      @features = 1.0
      @revenue = 1000
      @num_users = 0
      @turns_since_last_crash = 0
    end

    attr_reader :revenue, :num_users

    def will_crash_at(roll)
      get_fir > roll && num_users > 0
    end

    def get_fir
      @features.to_f / @infrastructure
    end

    def get_chance_of_crash   # view logic
      (get_fir * 100).round
    end

    def get_new_features_cost
      (500 * get_fir).round
    end

    def get_infrastructure_improvement_cost
      [800, @infrastructure / 3].max
    end

    def get_crash_cost
      exposure = [1.0, num_users / 10000.0].min
      (exposure * num_users * get_fir).to_i
    end

    def get_new_users(added_features)
      buzz_draw = added_features ? 1000 : 100
      (buzz_draw * @features * get_user_comfort_factor).to_i
    end

    def get_new_revenue(new_users)  # new_users is int
      ((@num_users - new_users * 0.75) + (new_users * 2)).to_i
    end

    def add_new_features
      @features += 1
      @revenue -= get_new_features_cost
    end

    def update_infrastructure
      @infrastructure += 1
      @revenue -= get_infrastructure_improvement_cost
    end

    def increment_turns_since_last_crash
      @turns_since_last_crash += 1
    end

    def reset_turns_since_last_crash
      @turns_since_last_crash = -1
    end

    def add_users(new_users)
      @num_users += new_users
    end

    def add_revenue(new_revenue)
      @revenue += new_revenue
    end

    private

    def get_user_comfort_factor
      [1 + @turns_since_last_crash, 10].max / 10
    end
  end
end
